### ISSUE TITLE
hls preload time and TTFB(Time To First Byte)

### ISSUE THREAD

**Issue by a695649350** [MoePlayer/DPlayer#769]
When loading the HLS video stream, each .ts file takes a long waiting time. Seams like each request is independent,and is there any way to request multiple files at one time to reduce waiting time?or increase the preload time
![1](https://user-images.githubusercontent.com/36436103/76683177-fc964980-663c-11ea-9c03-39d2ce4c4512.png)

**Comment by RexSkz** [comment]
DPlayer uses [video-dev/hls.js](https://github.com/video-dev/hls.js) to support HLS, which does not support parallel fetching, please refer to the following two issues:

- https://github.com/video-dev/hls.js/issues/2162
- https://github.com/video-dev/hls.js/issues/2536

However, you can optimize your server by using HTTP/2 or adding [Keep-Alive headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive) in file response, it can reuse the existing connection and reduce the TTFB.

**Comment by a695649350** [comment]
> DPlayer uses [video-dev/hls.js](https://github.com/video-dev/hls.js) to support HLS, which does not support parallel fetching, please refer to the following two issues:
> 
> * [video-dev/hls.js#2162](https://github.com/video-dev/hls.js/issues/2162)
> * [video-dev/hls.js#2536](https://github.com/video-dev/hls.js/issues/2536)
> 
> However, you can optimize your server by using HTTP/2 or adding [Keep-Alive headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive) in file response, it can reuse the existing connection and reduce the TTFB.

Thanks for reply.This isuuse happened in HTTP/2.I use nginx to add header and it seems like the Connection header will disappear in Http/2 as the HTTP document says. But how dose some file request multiple file as once(like js,css and jpg) when loading the static website source.

I am stucked in the trap that if I use big hls fragment the ratio of waiting time will decrease but download time will rise up,and vice versa.

**Comment by RexSkz** [comment]
That's weird... I did a little research and found that HTTP/2 maybe not be recommended for HLS.

The first reason is that large static files are considered low priority - by browser, which means we won't be able to do optimizations here.

Secondly, [this case](https://superuser.com/questions/931389/http-2-has-much-higher-ttfb-than-http-1-1) shows that **even if you query multiple files at the same time, server using HTTP/2 still returns the file one by one**, and of course, high-priority file first...

As for increasing the preload time, the default `maxBufferLength` setting in hls.js is `30s`, which is already enough in most cases, increasing this value can't solve the problem.

In general, the delay for HLS can be more than 10 seconds, if you really need to reduce the delay, using HTTP-FLV is a better choice.

**Comment by a695649350** [comment]
> That's weird... I did a little research and found that HTTP/2 maybe not be recommended for HLS.
> 
> The first reason is that large static files are considered low priority - by browser, which means we won't be able to do optimizations here.
> 
> Secondly, [this case](https://superuser.com/questions/931389/http-2-has-much-higher-ttfb-than-http-1-1) shows that **even if you query multiple files at the same time, server using HTTP/2 still returns the file one by one**, and of course, high-priority file first...
> 
> As for increasing the preload time, the default `maxBufferLength` setting in hls.js is `30s`, which is already enough in most cases, increasing this value can't solve the problem.
> 
> In general, the delay for HLS can be more than 10 seconds, if you really need to reduce the delay, using HTTP-FLV is a better choice.

Yes,I realize that increase buffer size will not help. I need use m3u8 to Encrypt video,and I found that Apple will improve m3u8 in this issue,so hope there are some good ways to deal with it at that time.

找半天也没找到方案，服务器放的远果然不适合这样搞m3u8，看看下半年新标准怎么解决

**Comment by umi** [comment]
I have a similar problem.
I'm doing low latency hls live streaming with 500ms chunks.
Even when the distance from the server to the client is long enough and the line speed is sufficient, TTFB becomes an issue regardless of the Keep-Alive setting.
Supporting cmaf will improve it somewhat, but it's not an essential solution.

**Comment by umi** [comment]
I have a similar problem.
I'm doing low latency hls live streaming with 500ms chunks.
If the distance from the server to the client is too far, even if the line speed is sufficient, TTFB becomes a problem regardless of the Keep-Alive setting.
Supporting cmaf will improve it somewhat, but it's not an essential solution.

**Comment by a695649350** [comment]
> I have a similar problem.
> I'm doing low latency hls live streaming with 500ms chunks.
> If the distance from the server to the client is too far, even if the line speed is sufficient, TTFB becomes a problem regardless of the Keep-Alive setting.
> Supporting cmaf will improve it somewhat, but it's not an essential solution.

https://developer.apple.com/videos/play/wwdc2019/502/
https://developer.apple.com/documentation/http_live_streaming/protocol_extension_for_low-latency_hls_preliminary_specification

I dont kown if the "EXT-X-PRELOAD-HINT" tag will solve this problem as Apple says that it can "Eliminating unnecessary round trips".

The waiting time should be twice the physical distance plus network fluctuations.If the waiting time plus the download time exceeds the time of each ts file,it will break up the video play.
